crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs

```diff
@@ -0,0 +1,258 @@
+///
+/// ## Example
+///
+/// ```python
+/// from typing import TypeVar
+///
+/// T = TypeVar("T")
+///
+///
+/// class GenericClass(Generic[T]): ...
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// class GenericClass[T]: ...
+/// ```
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+#[derive(ViolationMetadata)]
+pub(crate) struct NonPEP695TypeParameter {
+    name: String,
+    generic_kind: GenericKind,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum GenericKind {
+    GenericClass,
+    GenericFunction,
+}
+
+impl Violation for NonPEP695TypeParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Always;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NonPEP695TypeParameter { name, generic_kind } = self;
+        match generic_kind {
+            GenericKind::GenericClass => {
+                format!("Generic class `{name}` uses `Generic` subclass instead of type parameters")
+            }
+            GenericKind::GenericFunction => {
+                format!("Generic function `{name}` should use type parameters")
+            }
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use type parameters".to_string())
+    }
+}
+
+/// UP046
+pub(crate) fn non_pep695_generic_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    // PEP-695 syntax is only available on Python 3.12+
+    if checker.settings.target_version < PythonVersion::Py312 {
+        return;
+    }
+
+    let StmtClassDef {
+        name,
+        type_params,
+        arguments,
+        ..
+    } = class_def;
+
+    // it's a runtime error to mix type_params and Generic, so bail out early if we see existing
+    // type_params
+    if type_params.is_some() {
+        return;
+    }
+
+    let Some(arguments) = arguments.as_ref() else {
+        return;
+    };
+
+    // TODO(brent) only accept a single, Generic argument for now. I think it should be fine to have
+    // other arguments, but this simplifies the fix just to delete the argument list for now
+    let [Expr::Subscript(ExprSubscript {
+        value,
+        slice,
+        range,
+        ..
+    })] = arguments.args.as_ref()
+    else {
+        return;
+    };
+
+    if !checker.semantic().match_typing_expr(value, "Generic") {
+        return;
+    }
+
+    let vars = {
+        let mut visitor = TypeVarReferenceVisitor {
+            vars: vec![],
+            semantic: checker.semantic(),
+        };
+        visitor.visit_expr(slice);
+        visitor.vars
+    };
+
+    // Type variables must be unique; filter while preserving order.
+    let nvars = vars.len();
+    let type_vars = vars
+        .into_iter()
+        .unique_by(|TypeVar { name, .. }| name.id.as_str())
+        .collect::<Vec<_>>();
+
+    if type_vars.is_empty() {
+        return;
+    }
+
+    // non-unique type variables are runtime errors, so just bail out here
+    if type_vars.len() < nvars {
+        return;
+    }
```

Comment:
nit: to me it makes sense to check the typevars are not empty before filtering out duplicates

```suggestion
    if vars.is_empty() {
        return;
    }

    let type_vars: Vec<&str> = vars
        .iter()
        .unique_by(|tvar| &tvar.name.id)
        .collect();

    // non-unique type variables are runtime errors, so just bail out here
    if type_vars.len() < nvars.len() {
        return;
    }
```

Reply:
This section was bothering me too. I factored it out into a `check_type_vars` function to share it between the class and function rules and applied your suggestion here.